// @moralisweb3/streams-typings/usagestats.ts
interface totalsStats {
  /**
   * The total amount of webhooks delivered across all streams
   */
  totalWebhooksDelivered: number;
  /**
   * The total amount of failed webhooks across all streams
   */
  totalWebhooksFailed: number;
  /**
   * The total amount of logs processed across all streams, this includes failed webhooks
   */
  totalLogsProcessed: number;
  /**
   * The total amount of txs processed across all streams, this includes failed webhooks
   */
  totalTxsProcessed: number;
  /**
   * The total amount of internal txs processed across all streams, this includes failed webhooks
   */
  totalTxsInternalProcessed: number;
}

interface UsageStatsStreams extends totalsStats {
  /**
   * The stream id
   */
  streamId: string;
}

export interface UsageStatsModel extends totalsStats {
  /**
   * Array of stream stats
   */
  streams?: UsageStatsStreams[];
  /**
   * The date since this stats are being counted
   */
  createdAt?: Date;
  /**
   * The date since this stats were last updated
   */
  updatedAt?: Date;
}
